controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -138,10 +144,15 @@ public void processingFailure(Exception error) {
                 log.error("truncateSegment {}/{}/{} error", scope, stream, segmentNumber, error);
                 result.completeExceptionally(error);
             }
+
+            @Override
+            public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenCheckFailed) {
+                result.completeExceptionally(new PravegaAuthenticationException(authTokenCheckFailed.toString()));
```

Comment:
this exception seems problematic.. 
is this retryable? if so, we should add the retryable trait to this exception so that its retried automatically.

also, there should be some mechanism with which we renew the token if we get auto failed exception back from segment store. 

so i suggest we do two things:
1. make authFailed a retryable exception (add the retryable trait)
2. have a cache for the token.. instead of passing the token around from top layers, let segment helper actually retrieve the token from the cache.. 
this would be a loadingcache and on a cache miss it will renew the token.. 
it will also optimistically look for opportunities to renew the token during getToken calls. 

also the cache should be invalidated if we get auth-failed exception. so that during a retry we retrieve new token..

Reply:
1. In my opinion `AuthenticationException` should not be a retriable exception. Though this makes sense on the controller side, for a generic client this is an important failure which needs to be passed on to the app immediately. There are two possibilities. a. The controller throws the exception: This is passed on to the app. b. Segmentstore throws the exception:On segmentstore side, it is secure to drop the connection once we see this error. The client visits the controller in case of connection dropped error to refresh the metadata. During this call even the auth token will be refreshed. This takes care of withdrawn creds. So this is already covered.
2. Delegation token caching is a generic TBD issue. I will log a github issue for this. I do not think we should solve it specifically for controller. It should be solved for all the clients.

Reply:
Then we should have a separate retryable-exception on the controller which ensures that in case of auth failures we initiate a retry. 
So in segment helper, we should complete our futures with that controller specific exception. 

the problem is critical for controller because we dont want to have workflows fail in a partially completed state. 
and workflows are retried if and only if there is a retryable exception.

Reply:
Made this exception an retryable exception as suggested.